test/integration/filters/set_response_code_filter.cc

```diff
@@ -0,0 +1,57 @@
+#include <string>
+
+#include "envoy/http/filter.h"
+#include "envoy/registry/registry.h"
+
+#include "extensions/filters/http/common/factory_base.h"
+#include "extensions/filters/http/common/pass_through_filter.h"
+
+#include "test/integration/filters/set_response_code_filter_config.pb.h"
+#include "test/integration/filters/set_response_code_filter_config.pb.validate.h"
+
+#include "absl/strings/match.h"
+
+namespace Envoy {
+
+// A test filter that responds directly with a code on a prefix match.
+class SetResponseCodeFilter : public Http::PassThroughFilter {
+public:
+  SetResponseCodeFilter(const std::string& prefix, uint32_t code,
+                        Server::Configuration::FactoryContext& context)
+      : prefix_(prefix), code_(code), tls_slot_(context.threadLocal().allocateSlot()) {}
+
+  Http::FilterHeadersStatus decodeHeaders(Http::RequestHeaderMap& headers, bool) override {
+    if (absl::StartsWith(headers.Path()->value().getStringView(), prefix_)) {
+      decoder_callbacks_->sendLocalReply(static_cast<Http::Code>(code_), "", nullptr, absl::nullopt,
+                                         "");
+      return Http::FilterHeadersStatus::StopIteration;
+    }
+    return Http::FilterHeadersStatus::Continue;
+  }
+
+private:
+  const std::string prefix_;
+  const uint32_t code_;
+  // Allocate a slot to validate that it is destroyed on a main thread only.
+  ThreadLocal::SlotPtr tls_slot_;
+};
+
+class SetResponseCodeFilterFactory : public Extensions::HttpFilters::Common::FactoryBase<
+                                         test::integration::filters::SetResponseCodeFilterConfig> {
+public:
+  SetResponseCodeFilterFactory() : FactoryBase("set-response-code-filter") {}
+
+private:
+  Http::FilterFactoryCb createFilterFactoryFromProtoTyped(
+      const test::integration::filters::SetResponseCodeFilterConfig& proto_config,
+      const std::string&, Server::Configuration::FactoryContext& context) override {
+    auto filter_config = std::make_shared<SetResponseCodeFilter>(proto_config.prefix(),
+                                                                 proto_config.code(), context);
```

Comment:
nit: can we make this more realistic. The filter should be allocated inside the lambda below, with the TLS slot allocated outside. This could be done by just allocating the slot outside and passing it, or have a config object and passing that to the filter as is done elsewhere.

Reply:
I was just copying what wasm filter did. I guess it's not idiomatic.